V8: Ability to add a custom validation feedback (#5702)

AndyButland opened the pull request:
This PR is updated from [this closed one](https://github.com/umbraco/Umbraco-CMS/pull/1479) for V7 and addresses the new created issue #5702.

It adds two new fields that can be completed for each property type on a document type, that will be displayed if on content entry validation fails for the mandatory setting, or the regular expression based validation.

![image](https://user-images.githubusercontent.com/1993459/60272238-126baf00-98f4-11e9-9c49-e0e7502c5a79.png)

If completed, these custom messages will be displayed when validation fails, in preference to the default messages.

![image](https://user-images.githubusercontent.com/1993459/60272310-30d1aa80-98f4-11e9-8dc1-665528aafb2b.png)

In working on this I found a number of other issues related to validation, but have verified they exist with the latest `v8/dev` branch so aren't something I've introduced here.  They are likely best handled (or may already exist) as other issues.  For reference they are:

- A mandatory radio-button list colour picker reports that it's not valid even when a value is selected.
- A mandatory check-box list doesn't require one value to be checked
- When a regular expression validation fails, the invalid value is still saved

---
_This item has been added to our backlog [AB#3408](https://umbraco.visualstudio.com/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/3408)_

poornimanayar commented:
Hi @AndyButland ,

Great work! I tested the v7 one and loved the functionality. Amazing that you managed to do this PR for v8. If you can resolve the conflicts here, that would be great!

Poornima

AndyButland commented:
Conflicts should be resolved now @poornimanayar.

poornimanayar commented:
Hi @AndyButland ,

Is this SQL still valid and needs to be run for this PR?

```
ALTER TABLE dbo.cmsPropertyType ADD
    mandatoryMessage nvarchar(500) NULL,
    validationRegExpMessage nvarchar(500) NULL
GO
```

Poornima

AndyButland commented:
No, there's a migration in the PR that'll create those two columns.

Expect you've come across this already, but if you checkout and build to test this PR, and run the migration, if you then want to check out `v8/dev` again and run that, you'll need to update the database so it doesn't think this migration has been run.  At the time of writing, you can do that with:

```
update umbracoKeyValue set value = '{5B1E0D93-F5A3-449B-84BA-65366B84E2D4}'
where [key] = 'Umbraco.Core.Upgrader.State+Umbraco.Core'
```

Where the Guid is the latest migration noted in `UmbracoPlan.cs`.

---

BTW, just noticed the failing test above... will take a look at that a bit later.

poornimanayar commented:
Hi @AndyButland ,

Yes, I am getting an error on front-end build as well.

Poornima

AndyButland commented:
Have resolved the failing unit test.  There may well be a better way to do this but I'm not that familiar with it, so have just added a null check.

poornimanayar commented:
Hi @AndyButland ,

I have started testing this and found some issues around the reqd error messages
a) A mandatory textarea doesnt validate at all, it could be another issue related to the validation itself. I shall give it a test on v8/dev
b) On most of the datatypes I tested the validation message shows at the bottom with the "Property has errors" banner above it. But its different for colour picker alone where the custom error message shows in the banner. See an example below
![image](https://user-images.githubusercontent.com/3941753/60385137-d82e1900-9a7d-11e9-8f50-d8d616545b78.png)

poornimanayar commented:
Hi @AndyButland ,

So I have checked `v8/dev` and i can see that both the issues i mentioned above can be replicated. So I shall continue testing.

Poornima

poornimanayar commented:
Hi @AndyButland ,

Some more things I found. 

1. Textarea is validated upon save & publish
2. banner is showing the error message on the textarea validation
3. numeric, when mandatory, always has errors 

All these are issues with validation. I have tested textstring, textarea, dropdown, radiobuttonlist, checkboxlist, numeric and colour picker. Is there anything more I need to test?

For me this PR works, its the inconsistencies in the validation which is already present in the core which makes things a bit confusing. Its a great feature and great work! 

@nul800sebastiaan can you cast an eye on this when you have a moment?

Poornima

AndyButland commented:
Thanks for checking it over @poornimanayar - there's certainly a few existing issues around validation, as you've seen, and I've noted a few more in the initial comment on this PR.  I'm pretty sure as you've found they are in the existing code-base, so adding this feature isn't introducing anything problematic.  But they all likely need looking at via separate issues and PRs.

Shazwazza commented:
Just a quick note since i saw this mentioned as a bug a couple times and i think on the issue to - If server side validation fails on a property it is intended that this value gets saved. However, if the user is trying to publish the changes and server side validation fails, the content item will still be saved but it will not be published. That is perfectly normal and is what we want to have happen. 
I realize that this can be confusing sometimes because when client side validation fails the values are not persisted, and that just because that's how client side validation works and prevents the form from being submitted to the server in the first place.

AndyButland commented:
Thanks @Shazwazza - yes, I realised after a bit of digging that this was by design.  It was a little confusing in use for the reason you say, and also that you get the "you have unsaved changes" message pop-up still, even though the values have been saved.

Not really for this PR, but might be worth considering if the behaviour is actually what is best for invalid fields.  I.e. I can see the reasons for it for required fields - perhaps you've got a lot to fill out, and want to save part-way through.  But for an invalid entry failing the regular expression, it's not so clear it's useful to save that.

nielslyngsoe commented:
Hi @AndyButland sorry for the late follow up. I think this is great, so let's get it merged in soon. Its does need to get up to date with latest stuff from 'v8/dev', when thats in place let me know and I will review again and merge.

I do have a minor correction if you don't mind; how would it work for you if the inputs was the type <input type="text"> instead of <textarea>? That would nudge people to write short and precise messages on a single line.

AndyButland commented:
Hi @nielslyngsoe - I've updated by merging in `v8/dev` and resolved conflicts.  Have also switched to single-line input fields as requested, and moved the database migration into the update into version 8.3.0, which as of now looks to be the appropriate one.

Shazwazza commented on src/Umbraco.Web.UI.Client/src/common/services/validationhelper.service.js:
```diff
@@ -0,0 +1,34 @@
+(function () {
+    'use strict';
+
+    function validationHelperService($q, localizationService) {
```
We are trying to avoid the term `helper` in our code. We have quite a lot of code dealing with validation and `validationHelper` is too generic. Is there a different/more specific name for this service we could use? Or can this method exist on an already existing service?

AndyButland replied:
Updated to a more specific `validationMessageService`.

Shazwazza replied:
Small thing - but the file name will need to be change too,

AndyButland replied:
Updated.

AndyButland commented on src/Umbraco.Core/PropertyEditors/IDataValueEditor.cs:
```diff
@@ -35,17 +35,20 @@ public interface IDataValueEditor
         /// <param name="format">A specific format (regex) that the property value must respect.</param>
-        IEnumerable<ValidationResult> Validate(object value, bool required, string format);
+        /// <param name="formatMessage">A custom validation message to use when the property value is does not match the specific format (regex).</param>
+        IEnumerable<ValidationResult> Validate(object value, bool required, string requiredMessage, string format, string formatMessage);
```
I'll have a think... though I did notice in getting this PR up to date, that this interface was changed with the addition of `CanCopy` property about 3 weeks ago, so this breaking change issue may existing with minor upgrades already.

Would adding a new method (i.e. an overload) be better than changing the existing signature do you think?  Or is that still as much of a concern?

Shazwazza replied:
Hi @AndyButland i think that change somehow slipped through the cracks, we're gonna have to adjust that somehow before the next release too so there's no breaks before 8.2.

The problem with not changing the interface is that you are then gonna have to do a type check on the concrete type and then call that method which is a little uglier than having a version 2 of the interface since at least that allows you to mock stuff. Another alternative is to have just another standalone interface containing the changes but that's also not exactly nice.

We'll have a chat about these changes, perhaps we can allow breaks for this in 8.3 but will let you know soon :)

AndyButland replied:
OK, will leave for now then - as if you take away the breaking changes issue, then of course it's cleaner as is, to just change the interface.  But if that's not feasible let me know, probably along with which of the suggestions you've made you think best, and I can look to amend.

Shazwazza replied:
Hi @AndyButland , I 'think' i might have a way to acheive this without breaking changes. 

The underlying mechanism that calls the validation is only in 2 places: 

* ContentModelValidator.ValidatePropertyValue
* PropertyValidationService.IsPropertyValueValid

I 'think' in these 2 places we could call into `IEnumerable<ValidationResult> Validate(object value, bool required, string format);` 2 different times depending on the values there are for `required` and `format`. So first we would check if the property type is marked required, if so, we would call this method with `required: true, format: null` and get the response, if it was invalid then we can assign the property type's required message to the outgoing validation. Then we could call this method again if the property type had a regex format applied and call the method like `required: false, format: format` and do the same again, if the result was invalid we'd assign the property type's format message to the outgoing validation. 

From what i can tell this should be possible and then none of these interfaces need to change, it is then just the responsibility of the callers to the `Validate(object value, bool required, string format);` method to either use the error message returned or replace it with a configured one in the property type.

Does that make sense? Maybe it's a little ugly but this could then all be wrapped in an extension method on `IDataValueEditor`

AndyButland replied:
Thanks for the pointers, I'll take a look later on and see if I can make some progress.

AndyButland replied:
Unfortunately I ran into an issue with this @Shazwazza  - which is that the implementation of `DataValidator.Validate` will check not just required and regex validation but also any custom validators that are associated with the property type, `UploadFileTypeValidator` being one example.

So there actually isn't a way, with the current method signature, to tell it to just do required validation, or regex validation... we'll do one or other of those, but also any custom validations too.  And as such any `ValidationResult` objects in the collection that's returned, may be from the required/regex validation we've requested, or may come from the custom validators.  We'll also do the custom validators twice, potentially ending up with duplicate validation errors.

So I tried a different tack.  Given we know what the default message applied will be, we can just look for the cases where we have a `ValidationResult` with those error messages, and replace them.

Seems we actually only need this in `ContentModelValidator`.  The other instance you noted, in `PropertyValidationService` is just interested in a valid or not boolean results, so we don't care about messages there.  Given that, didn't see there was really a need to pull this out into an extension method, it can just live in `ContentModelValidator`.

So I've done that in the latest commit - as well as reverting the interface changes.  Please let me know how that looks to you.  It's still a little ugly I think relying on matching error messages - and I suppose there's a very small edge case in a custom validator happening to use the same message string -  but it looks to do the job.

Shazwazza replied:
Thanks @AndyButland ! It is a bit ugly so i guess now we need to decide: Do we allow breaking changes and so this the right way or do we do this the hacky way. There are some other upcoming changes to property editors with regards to media tracking that may end up resulting in a requirement to change interfaces (not decided yet), so maybe if that happens we can align these breaking changes in the same release. 

I have discovered that we are already (unfortunately) casting `IDataValueEditor` -> `DataValueEditor` 100% of the time within `DataEditor.GetValueEditor` and there is a note: `// TODO: casting is bad` which i assume used to be a `fixme` meaning we probably didn't want to ship with this. So in theory, the `Configuration` should be part of the interface too so we don't cast. 

Another alternative to what you've done would be to:

* Obsolete the `IDataValueEditor.Validate` method ... it could be left as-is though
* Update the `ValidatePropertyValue` method:
```cs
protected virtual void ValidatePropertyValue(
    TModelSave model,
    TModelWithProperties modelWithProperties,
    IDataEditor editor,
    ContentPropertyDto property,
    object postedValue,
    ModelStateDictionary modelState)
{
    // validate
    var valueEditor = editor.GetValueEditor(property.DataType.Configuration);

    // validate each custom validator
    List<ValidationResult> results = null;
    var r = valueEditor.Validators.SelectMany(v => v.Validate(postedValue, valueEditor.ValueType, property.DataType.Configuration)).ToList();
    if (r.Any()) { results = r; }

    // TODO: Casting is bad :(
    if (valueEditor is DataValueEditor dataValueEditor)
    {
        // validate required/regex
        if (property.IsRequired)
        {
            r = dataValueEditor.RequiredValidator.ValidateRequired(postedValue, valueEditor.ValueType).ToList();
            if (r.Any()) { if (results == null) results = r; else results.AddRange(r); }
        }

        var stringValue = postedValue?.ToString();
        if (!string.IsNullOrWhiteSpace(property.ValidationRegExp) && !string.IsNullOrWhiteSpace(stringValue))
        {
            r = dataValueEditor.FormatValidator.ValidateFormat(postedValue, valueEditor.ValueType, property.ValidationRegExp).ToList();
            if (r.Any()) { if (results == null) results = r; else results.AddRange(r); }
        }
    }
            
    foreach (var valResult in results)
    {
        modelState.AddPropertyError(valResult, property.Alias, property.Culture);
    }
}
```

Essentially just moving the full validation logic here which includes checking/casting which isn't nice but i think it might be nicer for now than checking validation message responses. Then from there we can decide when/how to break the interfaces to do all of this correctly. 

What do you think?

AndyButland replied:
I'm not quite following where in this last suggestion we'd be setting the custom messages though?  As if we do that in `RequiredValidator` and `RegexValidator`, we'd need to pass in the message to use, which would involve a change to `IValueRequiredValidator` and `IValueFormatValidator` interfaces.  And if we do it in `ValidatePropertyValue`, then seems we have to do the checking for matches on the error messages still.

Shazwazza replied:
@AndyButland ah yes of course since the result of ValidateRequired and ValidateFormat return multiple errors, we'd have to do the error message check anyways. Lets stick to your current solution for now, maybe that's just the simplest way to do it for now. Perhaps we should `Contains` to `ContainsInvariant` though just to be safer? Also just want to confirm this would still work when using a non-default back office language?

AndyButland replied:
Have updated to use `StringComparer.OrdinalIgnoreCase`.  I haven't tested yet using a non-English back-office language (and am away for a few days, so won't be able to do so shortly), but I believe it would work as expected.  We're using the localisation text service to get the default validation messages, which uses the configured culture, so it means we'd have the localised ones set and used for the check.

AndyButland replied:
Confirmed this works OK when I use a different back-office language and ensure these default validation messages are added to the appropriate language file.

Shazwazza replied:
Great stuff, i think this is in a good place, i'll find out from the team which version we should release this in and get the ball rolling. Thanks @AndyButland !!

warrenbuckley commented:
Note to ourselves: We will need to move the migration classes as 8.3.0 is already out...

warrenbuckley commented:
@AndyButland can you please fix the merge conflicts for me, as I am testing this today and can get it merged in shortly.

Also see my note about the migration namespaces/classes if you can do that too that would be amazing 🎉

AndyButland commented:
I've merged in the latest and updated the migration references.  Am getting some FE build issues locally that I don't have time to resolve now I'm afraid, so haven't been able to re-run to check it's all working as it was before, but conflicts were minimal so think it should be OK.

warrenbuckley commented:
Thanks, Andy
I will take a look at this today and get this done. Thanks again
